rust/src/core.rs

```diff
@@ -65,12 +65,11 @@ impl std::fmt::Display for Direction {
 
 impl From<u8> for Direction {
     fn from(d: u8) -> Self {
+         // d contém DIR_TOSERVER união DIR_TOCLIENT
         if d & (DIR_TOSERVER | DIR_TOCLIENT) == (DIR_TOSERVER | DIR_TOCLIENT) {
             debug_validate_fail!("Both directions are set");
             Direction::ToServer
-        } else if d & DIR_TOSERVER != 0 {
-            Direction::ToServer
-        } else if d & DIR_TOCLIENT != 0 {
```

Comment:
I'm not really sure if this is the right fix. But, I'll let you and someone else assess this too.
in
```
        if d & (DIR_TOSERVER | DIR_TOCLIENT) == (DIR_TOSERVER | DIR_TOCLIENT) {
```
we check if both the directions are set. We do `0x04 | 0x08` which in binary is `00000100 | 00001000` == `00001100`. We get into this block iff at least these two bits are set in the `d` flag.
in
```
else if d & DIR_TOSERVER != 0
```
we get into the block iff at least 0x04 is set.
same as above for the next block.

I _think_ clippy is a bit confused because it doesn't recognize our `debug_validate_fail!` which happens to be internal to our code. I also think that clippy is highlighting a different part of code than it should. I would expect it to highlight the last `else` block with this issue...
But, I think @jasonish would be the best person to check this.

Please feel free to correct me if you think this is wrong.

Reply:
So clippy is confused, running clippy like this should help:
```
cargo clippy --features "debug debug-validate"
```
that will allow clippy to recognize that code that macro expands to.

We should always run clippy like this, well with all features enabled.  I'll try to make this more clear where applicable.

Reply:
Thanks for the tip. I didn't know about those parameters, do they appear in the documentation or a tutorial and I missed it ?

Reply:
> Thanks for the tip. I didn't know about those parameters, do they appear in the documentation or a tutorial and I missed it ?

No they didn't not.. Just catching some of this stuff as we go.